File: Lab-10-petclinic-solution/src/main/java/com/petclinic/core/OwnerController.java
package com.petclinic.core;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
public class OwnerController {
    private final OwnerService ownerService;

    public OwnerController(OwnerService ownerService) {
        this.ownerService = ownerService;
    }

    @RequestMapping("/owner/{id}")
    public Owner findById(@PathVariable Long id) {
        return this.ownerService.findById(id);
    }

    @RequestMapping("/owner/search")
    public Owner findByFirstName(@RequestParam String firstName) {
        return this.ownerService.findByFirstName(firstName);
    }

    @PostMapping("/owner")
    public ResponseEntity<Owner> save (@RequestBody Owner owner) {
        var savedOwner = this.ownerService.save(owner);
        return ResponseEntity.status(HttpStatus.CREATED).body(savedOwner);
    }
}

File: Lab-kotlin/src/main/kotlin/com/petclinic/core/OwnerController.kt
package com.petclinic.core

import io.swagger.v3.oas.annotations.Operation
import jakarta.persistence.EntityNotFoundException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/owner")
class OwnerController(
    val ownerService: OwnerService) {

    @Operation(summary = "search an owner by his/her first name")
    @GetMapping("/search")
    fun findByFirstName(@RequestParam firstName: String): Owner? {
        return ownerService.findByFirstName(firstName)
    }

    @GetMapping("/{id}")
    fun findVisit(@PathVariable id: Long): Owner {
        return ownerService.findById(id)
    }

    @PostMapping("/")
    fun save(@RequestBody owner: Owner): ResponseEntity<Owner> {
        val savedOwner = ownerService.save(owner)
        return ResponseEntity.status(HttpStatus.CREATED).body(savedOwner);
    }

    @ExceptionHandler(EntityNotFoundException::class)
    fun handleNoSuchElementException(ex: EntityNotFoundException): ResponseEntity<Any> {
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Error ${HttpStatus.NOT_FOUND}. ${ex.message}")
    }
}